### ISSUE TITLE
[2.7] test_ttk_guionly doesn't destroy all widgets on Python 2.7

### ISSUE THREAD

**Issue by vstinner** [python/cpython#75318]
BPO | [31135](https://bugs.python.org/issue31135)
--- | :---
Nosy | @vstinner, @serhiy-storchaka
PRs | <li>python/cpython#3025</li><li>python/cpython#3026</li><li>python/cpython#3030</li>
Files | <li>[widget_instance.patch](https://bugs.python.org/file47061/widget_instance.patch "Uploaded as text/plain at 2017-08-08.00:46:37 by @vstinner")</li><li>[fix_destroy.patch](https://bugs.python.org/file47062/fix_destroy.patch "Uploaded as text/plain at 2017-08-08.00:46:59 by @vstinner")</li><li>[widget_instance-master.patch](https://bugs.python.org/file47063/widget_instance-master.patch "Uploaded as text/plain at 2017-08-08.12:41:39 by @vstinner")</li><li>[widget_instance-2.7-v2.patch](https://bugs.python.org/file47064/widget_instance-2.7-v2.patch "Uploaded as text/plain at 2017-08-08.13:04:49 by @vstinner")</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = <Date 2017-08-10.15:03:03.647>
created_at = <Date 2017-08-08.00:46:39.117>
labels = ['3.7', 'expert-tkinter', 'performance']
title = "[2.7] test_ttk_guionly doesn't destroy all widgets on Python 2.7"
updated_at = <Date 2017-08-10.15:03:03.644>
user = 'https://github.com/vstinner'
```

bugs.python.org fields:
```python
activity = <Date 2017-08-10.15:03:03.644>
actor = 'vstinner'
assignee = 'none'
closed = True
closed_date = <Date 2017-08-10.15:03:03.647>
closer = 'vstinner'
components = ['Tkinter']
creation = <Date 2017-08-08.00:46:39.117>
creator = 'vstinner'
dependencies = []
files = ['47061', '47062', '47063', '47064']
hgrepos = []
issue_num = 31135
keywords = ['patch']
message_count = 12.0
messages = ['299880', '299881', '299899', '299902', '299913', '299914', '299915', '299916', '299938', '299940', '299977', '300097']
nosy_count = 3.0
nosy_names = ['vstinner', 'gpolo', 'serhiy.storchaka']
pr_nums = ['3025', '3026', '3030']
priority = 'normal'
resolution = 'fixed'
stage = 'resolved'
status = 'closed'
superseder = None
type = 'resource usage'
url = 'https://bugs.python.org/issue31135'
versions = ['Python 2.7', 'Python 3.6', 'Python 3.7']
```

</p></details>

**Comment by vstinner** [comment]
While working on bpo-31068, I wrote attached widget_instance.patch to check if test_ttk_guionly really removes all widgets. Bad news: it doesn't. Example:

haypo@selma$ ./python -m test  -u all test_ttk_guionly
(...)
test test_ttk_guionly crashed -- \<type 'exceptions.Exception'\>: leaking 69 widgets
(...)

Attached fix_destroy.patch is my attempt to fix this issue.

**Comment by vstinner** [comment]
I only tested Python 2.7. I didn't check if other branches are affected.

**Comment by serhiy-storchaka** [comment]
See bpo-25130. Maybe forcing garbage collecting will help.

**Comment by vstinner** [comment]
My widget_instance.patch calls test_support.gc_collect() and removes all
widgets from the root window. It's not enough. That's why I proposed to modify LabeledScale.destroy().

**Comment by vstinner** [comment]
On master, LabeledScaleTest.test_initialization() leaks a widget in the root widget, even after root.destroy(). The problem is that LabeledScaleTest.\_\_del__() doesn't call Frame.destroy(self) if self.\_variable is not set. So master has at least this bug.

**Comment by vstinner** [comment]
widget_instance-master.patch: Check that Tkinter.Widget.destroy() removes all children and check that test_ttk_guionly doesn't leak widgets.

**Comment by vstinner** [comment]
Example of failure on master using widget_instance-master.patch:

haypo@selma$ ./python -m test -v -u all test_ttk_guionly 

\======================================================================
ERROR: test_initialization (tkinter.test.test_ttk.test_extensions.LabeledScaleTest)
\----------------------------------------------------------------------
```pytb
Traceback (most recent call last):
  File "/home/haypo/prog/python/master/Lib/tkinter/test/test_ttk/test_extensions.py", line 14, in tearDown
    super().tearDown()
  File "/home/haypo/prog/python/master/Lib/tkinter/test/support.py", line 40, in tearDown
    w.destroy()
  File "/home/haypo/prog/python/master/Lib/tkinter/__init__.py", line 2305, in destroy
    % self.children)
Exception: destroy() doesn't clear all children: {'!labeledscale2': <tkinter.ttk.LabeledScale object .!frame.!labeledscale2>}
```



My https://github.com/python/cpython/pull/3025 PR fixes LabeledScale and OptionMenu destroy() method in the master branch.

**Comment by vstinner** [comment]
widget_instance-2.7-v2.patch: Updated patch for Python 2.7 to debug test_ttk_guionly.

https://github.com/python/cpython/pull/3026 fix LabeledScale and OptionMenu destroy() method for Python 2.7.

**Comment by vstinner** [comment]
New changeset 56e162ad5c5d3effe9b4f05d0179e1b6a2a2d9b8 by Victor Stinner in branch '2.7':
ttk: fix LabeledScale and OptionMenu destroy() method (bpo-3026)
https://github.com/python/cpython/commit/56e162ad5c5d3effe9b4f05d0179e1b6a2a2d9b8

**Comment by vstinner** [comment]
New changeset cd7e9c1b67d3d07ee03e0a79af2791c19031cecb by Victor Stinner in branch 'master':
ttk: fix LabeledScale and OptionMenu destroy() method (bpo-3025)
https://github.com/python/cpython/commit/cd7e9c1b67d3d07ee03e0a79af2791c19031cecb

**Comment by vstinner** [comment]
New changeset 33460fa7e0bd126bee739a66e1228665dc22e70f by Victor Stinner in branch '3.6':
ttk: fix LabeledScale and OptionMenu destroy() method (bpo-3025) (bpo-3030)
https://github.com/python/cpython/commit/33460fa7e0bd126bee739a66e1228665dc22e70f

**Comment by vstinner** [comment]
Copy of interesting comments: https://github.com/python/cpython/pull/3025

serhiy-storchaka: "Parent's destroy() now is called even if this destroy() already was called. I.e. it can be called twice."

haypo: "Yes, it's a deliberate choice. All other ttk widgets now have the same behaviour."

--

Ok, the bug is now fixed in 2.7, 3.6 and 3.7 (master) branches. I close the issue.
